[microcosm] Unify effect and domain behavior through agent class

nhunzaker opened the pull request:
This commit adds a new class to describe the dispatch behavior: an agent. Agents receive action dispatches from history and can manage their own internal state:

```javascript
import { start, stop } from '../actions/gps'

class GPS extends Agent () {
  setup() {
    this.listen()
  }

  listen() {
    this.watcher = navigator.geolocation.watchPosition(this.next)
  }

  ignore() {
    navigator.geolocation.clearWatch(this.watcher);
  }

  receive(action) {
    switch(action) {
      case start: 
        return this.listen()
      case stop: 
        return this.ignore()
      default: 
        return null
    }
  }
}

let gps = new GPS()

gps.subscribe(next => console.log(next)) // location.... location .... location....

gps.send(start)
```

I still haven't figured out the public API for Agents yet, but I think they will be a useful escape hatch for creating new types of Domain/Effect-like abstractions. I'm also curious if they could be used as a type of viewmodel class, but I'm not quite there yet.

nhunzaker commented on packages/microcosm/src/subject.js:
```diff
@@ -6,14 +6,14 @@ import { noop, EMPTY_SUBSCRIPTION } from './empty'
 
 export class Subject {
-  meta: { tag: *, status: string, origin: Microcosm }
+  meta: { key: *, status: string, origin: Microcosm }
```
I renamed `tag` to `key` for consistencies sake.

nhunzaker commented on packages/microcosm/src/subject.js:
```diff
@@ -134,14 +134,14 @@ export class Subject {
 
   toString(): string {
-    return this.meta.tag || 'Subject'
+    return this.meta.key
```
Feels good to get rid of this null state :)

nhunzaker commented on packages/microcosm/test/helpers.js:
```diff
@@ -4,7 +4,7 @@ export function asTree(history) {
+    return children.length ? [focus.key, ...children] : focus.key
   }
 
   return asciitree(dig(history.toJSON().tree))
```
As a side note, this is a utility for serializing history into a tree visualization. I'm using this for snapshot testing the history structure:

```
    a
    |
    b
   / \
  c   d
```

Super useful!

nhunzaker commented on packages/microcosm/src/agent.js:
```diff
@@ -0,0 +1,52 @@
+
+  static defaults: ?Object
+
+  constructor(repo: Microcosm, options?: Object) {
```
Thinking on this a bit more, I'm also curious about the idea of an `AgentPool` class that could distribute state management across a group of presenters by sharing agents with the same configuration.

nhunzaker replied:
You'd do that by deriving a hashcode from the repo's identity and values passed via options

dce commented on packages/microcosm/src/domain.js:
```diff
@@ -116,10 +62,22 @@ export class Domain<State: mixed = null> extends Subject {
+    // that cleaned up keys as actions completed
+    this._ledger.clean(action)
+  }
+
```
❤️ how much this cleans up Domain.

dce commented on packages/microcosm/src/proto.js:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  return NewEntity
+}
```
Does this have a test?

nhunzaker replied:
It gets covered through domain/effect usage in other tests, but thanks for calling it out. I should test to cover a few cases:

- `null`
- There are certain rules for [property descriptors of ES6 classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptors), I should make this play nice

Really, I don't want to have something like this in Microcosm, but we've allowed users to pass domains/effects as POJOs forever. I'm thinking about deprecating it, but actually enforcing that domains/effects need to inherit from a library class might need to go out in 14.x.

efatsi commented on packages/microcosm/src/agent.js:
```diff
@@ -0,0 +1,52 @@
+      complete: () => {
+        tracker.unsubscribe()
+        this.teardown(this.repo, this.options)
+      }
```
looks like this is existing functionality from domains/effects, but what would cause a `complete` action?

nhunzaker replied:
Microcosm's are Observables (implemented as Subjects ). When you call `complete()` on a Microcosm, it has the same effect as calling `shutdown` in Microcosm 12. Instead of subscribing to the `teardown` event using an event emitter, repos/domains/presenters now listen to the `complete` lifecycle event.

TBH, I think there's an improvement that could be made here. This feels like bad design. I wonder if I should give Subject itself a setup and teardown method. You wouldn't have to implement it, but it could be inherited through Agent to Domains and Effects.

efatsi commented:
+1 to the refactoring here. Personally, I'm not too interested in a public API for Agents - I can't think of a use case for cutting off your own class that extends Agent when you have Domains/Effects at your fingertips.

nhunzaker commented:
@dce Looks like I was missing coverage in a few areas. In 1e37997a I added a few more tests and added an assertion to warn users when they add a null domain/effect. This achieves complete coverage of the proto module.

nhunzaker commented:
@efatsi Agreed. I don't want to advertise it for the time being, but I _am_ quite curious about how we could use it to create domains that do not rely on history.

nhunzaker commented:
Thanks everyone!
